### ISSUE TITLE
Write wav with generator input

### ISSUE THREAD

**Issue by djl0** [scipy/scipy#12429]
@endolith I'd love your input on the feasibility of this, and if I'm maybe missing a more elegant solution.

I'm wondering about passing wavfile.write() a generator of wav data, as opposed to all wav data. https://github.com/scipy/scipy/blob/e4238172b1925e1bf6a554678fd765553f0e069c/scipy/io/wavfile.py#L644

Reason:
The Mozilla TTS (text to speech) project uses wavfile.write. Generally it seems like this is writing a stream of data, just a phrase or 2 at a time. But when converting pages of text, holding that all in memory is unnecessary and impossible. usage: https://github.com/mozilla/TTS/blob/3366328126b329380dcf9d81b064976e9eb96e17/utils/audio.py#L320

I was poking around with it for a while, but I don't have much knowledge about wav headers the like. But do you think an elegant solution to this problem would be to be able to pass `data` variable as a generator? Then in the example above, text could be synthesized one sentence at a time, and then immediately dumped from memory.

Thank you for your input.

**Comment by endolith** [comment]
That should be possible, but you need to know the total length of data ahead of time to write the headers.  I guess you could write a malformed WAV with maximum length in the headers and it would probably still work in most software.  Or go back to the beginning of the file and modify the headers after the data is complete.

**Comment by djl0** [comment]
Thanks for your response! How difficult would it be to re-write/modify the header?

Also, would this work? Write main data portion of file with generator, keep track of length. Write header file at the end, and then in bash do: `cat main_data >> headers.wav`

**Comment by endolith** [comment]
I don't think it would be hard at all, but I'm thinking of regular PCM files with nothing but data chunks.  I would say to write the headers with 0xffffffff placeholders for the file and data chunk sizes, then go back and modify them after writing the rest. This becomes more complicated if other chunks are being written though, or maybe with other types of file-like objects.

**Comment by djl0** [comment]
I think I actually got it working using the same technique as the file-size write (determine final size, then seek(4), then write length-8). I create a 'bookmark' after writing `data` (though i think it's always going to be at `40`), and after summing of all `nbytes`, write the final byte size.

One other question for you: do you think this would be a useful refactor to have as part of the project? Or should this just be my own fork? I've never really made changes like this to an open source project before.

**Comment by endolith** [comment]
> do you think this would be a useful refactor to have as part of the project? Or should this just be my own fork?

I'm not sure.  I'd see what other people say.
